Add parent and children info to all Base nodes (for 102 and 105)

arporter opened the pull request.

arporter commented:
Having added `Base.get_root()` I realised that the `get_child()` routine ought to be a method of `Base` as well so I've made that change. Similarly, we should move the `walk_ast` routine to `Base.walk` but that's a bigger change so I've not done it in this PR.

arporter commented:
Ready for review now. Happy Christmas @rupertford :-)

arporter commented:
This PR could be viewed as a (small) step towards #105.

arporter commented:
I've just found another problem with this (I've omitted adding the `parent` argument to the `Comment.__new__` method) so taking it back.

arporter commented:
I've done it this way as it seemed the least intrusive. The problem is that most objects are created in the `match` method which is static and therefore doesn't have `self`. We therefore don't know what the parent is at the point the child objects are created. We could possibly change `match` so that it is no longer static but that felt like quite a big change so I opted for something simpler. Am happy to talk this through when we next see each other (or over Skype).

arporter commented:
(I've added the `parent` argument to the Base constructor to make it easier to construct a parse tree, e.g. when testing - it's not actually used during normal fparser operation.)

rupertford commented:
> I've done it this way as it seemed the least intrusive. The problem is that most objects are created in the `match` method which is static and therefore doesn't have `self`. We therefore don't know what the parent is at the point the child objects are created. We could possibly change `match` so that it is no longer static but that felt like quite a big change so I opted for something simpler. Am happy to talk this through when we next see each other (or over Skype).

I agree that match should stay static. Let's talk through when we're next in the office.

rupertford commented:
I know we agreed to wait but I had a quick play and I wonder whether the following would work if added to `Base` (where the *'ed code is the additional code). Note, I've not bothered with setting up parent (as a property or a variable).

```
        if isinstance(result, tuple):
            obj = object.__new__(cls)
            obj.string = string
            obj.item = None
            *from fparser.two import utils
            *for child in result:
            *    if isinstance(child, utils.Base):
            *        child.parent = obj
            if hasattr(cls, 'init'):
                obj.init(*result)
            return obj
```

arporter commented:
That looks like it might work. I hadn't quite grasped that `init` was only called from one place and that if we do the set-up before it then we only have to do it once which is clearly better. I'll have a play.

arporter commented on doc/developers_guide.rst:
```diff
@@ -315,9 +315,11 @@ such as programs, subroutines, if statements etc.).
 
 .. autoclass:: fparser.two.utils.BlockBase
 	       :members:
+               :noindex:
```
Done. Pesky tabs.

arporter commented:
In the process of addressing the review comments I decided life would be easier if I just went ahead and added a `children` method to `Base`. However, in https://github.com/stfc/fparser/issues/105#issuecomment-468690505, Phil suggested a visitor-style implementation of `children`.  This has the advantage that we can transparently support those nodes that are e.g. `str`. The same goes for the `parent` implementation - currently I have that as a property but that means that I have to check that a node is a sub-class of Base before attempting to get its parent. If this was done as a visitor we wouldn't have this problem. What do you think @rupertford?

arporter commented on src/fparser/two/Fortran2003.py:
```diff
@@ -103,7 +103,7 @@ class Comment(Base):
 
     @show_result
-    def __new__(cls, string, parent_cls=None):
+    def __new__(cls, string, parent_cls=None, parent=None):
```
Done.

arporter commented on doc/fparser2.rst:
```diff
@@ -392,10 +392,15 @@ file was found but would fail if the include file was not found::
 ----------------------
 
-fparser2 provides two functions to support the traversal of the
+fparser2 provides one utility function to support the traversal of the
```
Done.

arporter commented on doc/developers_guide.rst:
```diff
@@ -315,9 +315,11 @@ such as programs, subroutines, if statements etc.).
 
 .. autoclass:: fparser.two.utils.Base
 	       :members:
+	       :noindex:
```
Done.

arporter commented on doc/conf.py:
```diff
@@ -200,7 +200,7 @@
 # relative to this directory. They are copied after the builtin static files,
 # so a file named "default.css" will overwrite the builtin "default.css".
-html_static_path = ['_static']
+#html_static_path = ['_static']
```
Done.

arporter commented on doc/fparser2.rst:
```diff
@@ -392,10 +392,15 @@ file was found but would fail if the include file was not found::
 ----------------------
 
-fparser2 provides two functions to support the traversal of the
+fparser2 provides one utility function to support the traversal of the
```
Done.

arporter commented on doc/conf.py:
```diff
@@ -200,7 +200,7 @@
 # relative to this directory. They are copied after the builtin static files,
 # so a file named "default.css" will overwrite the builtin "default.css".
-html_static_path = ['_static']
+#html_static_path = ['_static']
```
It causes an error because the path doesn't exist. Space added.

rupertford replied:
OK, thanks for explaining.

arporter commented:
CodeCov has got itself confused and is claiming that docstring lines aren't covered! Apart from that I think this is ready for review again. I had to keep the `set_parent` method in order to allow for elements of the results list being lists/tuples. However, it is now only called from one location which is much better and I've made it a private routine so it is not a method on the `Base` class. I've implemented `get_children` and `get_parent` as visitors (if that's what they are) so you can see what that looks like and we can discuss :-)

arporter commented on src/fparser/two/tests/test_utils.py:
```diff
@@ -113,25 +113,59 @@ def test_blockbase_match_name_classes(f2003_create):
+    # Check connectivity of all non-string nodes
+    for node in node_list[1:]:
+        if isinstance(node, Base):
+            assert node.parent
```
Yes. Done.

rupertford commented on src/fparser/two/utils.py:
```diff
@@ -272,23 +272,52 @@ class Base(ComparableMixin):
     # 'subclass_names' list belonging to each class defined in this module.
     subclasses = {}
 
+    def __init__(self, string, parent_cls=None, parent=None):
```
As discussed, if the only reason for having `parent` being passed by argument is for testing purposes then I don't think it should be here.

arporter replied:
Removed.

rupertford commented on src/fparser/two/utils.py:
```diff
@@ -272,23 +272,52 @@ class Base(ComparableMixin):
+        # During normal construction of the parse tree the parent information
+        # is set-up in the `init` method. However, it may also be provided
+        # to this constructor.
+        self.parent = parent
```
This would obviously become `parent = None` if the argument is passed. An alternative would be to have `self._parent=None` and have a parent property to hide the implementation. I'm happy for you to choose which you prefer.

arporter replied:
I've gone for `self.parent = None`. I don't think hiding it behind a setter/getter buys us anything at the minute (and we can always change later without changing the interface).

rupertford commented on src/fparser/two/utils.py:
```diff
@@ -272,23 +272,52 @@ class Base(ComparableMixin):
-    def __new__(cls, string, parent_cls=None):
-        """
-        Create a new instance of this object.
+    def __new__(cls, string, parent_cls=None, parent=None):
```
Again, please remove `parent=None` if it is removed from `__init__`

arporter replied:
Done.

rupertford commented on src/fparser/two/utils.py:
```diff
@@ -335,6 +364,8 @@ def __new__(cls, string, parent_cls=None):
             obj.string = string
             obj.item = None
+            # Set-up parent information for the results of the match
+            _set_parent(obj, result)
```
As discussed, I think we should also have a list called `children` that is a member of `Base` and I think the simplest place to set it up would be here (perhaps _set_parent could become `_set_links` and do both?

arporter replied:
I've done this as a `children` property that transparently returns either `items` or `content`. Otherwise we would end up manually keeping a copy of either of those that we would have to keep up-to-date when the tree is manipulated.

rupertford commented on src/fparser/two/utils.py:
```diff
@@ -380,9 +411,45 @@ def __new__(cls, string, parent_cls=None):
             errmsg = u"{0}: '{1}'".format(cls.__name__, string)
         raise NoMatchError(errmsg)
 
+    def get_root(self):
```
I would change this to a `root` property to be consistent with parent and children, but would accept as is if you prefer it this way.

arporter replied:
I'm keeping it as get_root() as, as discussed, it could get costly to repeatedly call it if we're in a deep tree.

rupertford commented on src/fparser/two/utils.py:
```diff
@@ -380,9 +411,45 @@ def __new__(cls, string, parent_cls=None):
+        :rtype: :py:class:`fparser.two.utils.Base`
+
+        '''
+        parent = self
```
I would change the `parent` variable to `current`

arporter replied:
Good idea.

rupertford commented on src/fparser/two/utils.py:
```diff
@@ -380,9 +411,45 @@ def __new__(cls, string, parent_cls=None):
+
+        '''
+        parent = self
+        while parent.parent:
```
I suggest storing the result the first time to avoid iterating on each call.

arporter replied:
As discussed, this could cause problems when manipulating a parse tree so I'm leaving it as dynamic. I'll keep it as `get_root()` to signal to the user that there's some cost associated with it.

rupertford replied:
Agreed.

rupertford commented on src/fparser/two/utils.py:
```diff
@@ -380,9 +411,45 @@ def __new__(cls, string, parent_cls=None):
+            parent = parent.parent
+        return parent
+
+    def get_child(self, node_type):
```
I think we might have discussed removing this but I think we should keep it. Obviously `get_children(node)` would become `self.children` though and I would suggest the variable `node` is changed to `child`. Having said that, I would be happy for this method to be part of a separate utils package, along with walk, as it is not a fundamental part of the tree (unlike parent, children and root).

arporter replied:
I've changed `node` to `child` and kept this routine as part of the utils package. The `Base.get_child()` method has been deleted.

rupertford replied:
Great.

rupertford commented on src/fparser/two/utils.py:
```diff
@@ -629,11 +696,33 @@ def match(startcls, subclasses, endcls, reader,
         self.content = content
-        return
+
+    def get_child(self, node_type):
```
I think this can just be removed as we have a generic `get_child`?

arporter replied:
Done.

rupertford commented on src/fparser/two/utils.py:
```diff
@@ -1500,64 +1594,74 @@ def tostr(self):
 
 
-def get_child(root_node, node_type):
+def get_children(node):
```
As discussed, please replace this with a children property/attribute in the `Base` base class.

arporter replied:
Done.

rupertford commented on src/fparser/two/utils.py:
```diff
@@ -1500,64 +1594,74 @@ def tostr(self):
+    return child_list
+
+
+def get_parent(node):
```
As discussed, please remove this in favour of a parent property/attribute in the `Base` base class.

arporter replied:
Done.

arporter commented:
Hopefully I've made all the requested changes. Ready for next review...

arporter commented:
On reflection last night I realised that a user might expect to be able to manipulate the result obtained from using the new `children` property and that this would only work if the node was a statement and not an expression. I've therefore updated the property to return (the original) tuple of children in the latter case so that it is immutable by design. I've updated the docstring to explain this too.

arporter commented:
Tests added and some moved. Ready for next review.
